### PR TITLE
calculating losses along connection apths between plant and reservoir

### PR DESCRIPTION
**Pull request by Juliamg**
## Description
_Waiting to get logic approved by Øivind from Lyse._

UPDATE: Øivind has confirmed that logic looks right.

## Checklist:
- [ ] Tests added/updated.
- [ ] Documentation updated.
- [ ] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/power-ops-sdk/blob/master/CHANGELOG.md).
- [ ] Version bumped. If triggering a new release is desired, bump the version number in [pyproject.toml](https://github.com/cognitedata/power-ops-sdk/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: cognite/powerops/resync/models/_shared_v1_v2/production_model.py [hunk 1]
```diff
@@ -86,31 +84,70 @@ def _plant_to_inlet_reservoir_breadth_first_search(
...
+            if connection[-1] == last_visited_id:
+                connection_path_index = visited_paths.index(connection)
+                return visited_paths[connection_path_index]
+
+    def calculate_losses_from_conncetion_path(all_junctions, all_tunnels, connection_by_id, connection_path):
+        sum_losses = 0
+        order_to_loss_factor_key = {0: "loss_factor_1", 1: "loss_factor_2"}
+        for connection_id in connection_path:
+            connection = connection_by_id[connection_id]
+            if connection.get("to_type") == "junction":
```
Comment (Juliamg) [inline cognite/powerops/resync/models/_shared_v1_v2/production_model.py:99]: only include losses **to** a junction or tunnel

File: cognite/powerops/resync/config/production/plant.py [hunk 2]
```diff
@@ -26,6 +26,7 @@ class Plant(BaseModel):
     p_max: float = p_max_fallback  # arbitrary max power output if not specified
     head_loss_factor: float = 0.0
     penstock_head_loss_factors: Optional[dict[str, float]] = None
+    connections_losses: Optional[float] = None
```
Comment (Juliamg) [inline cognite/powerops/resync/config/production/plant.py:29]: Øivind asked if we could include potential conneciton losses in plant metadata so they could differentiate the plant that have and do not have connection losses

File: tests/test_unit/test_resync/approval_test/test_apply/demo_ProductionModel.yml [hunk 3]
```diff
@@ -180,8 +185,9 @@ assets:
   labels:
   - externalId: plant
   metadata:
+    connection_losses: '0.00231'
     display_name: Rull1
-    head_loss_factor: '0.00107'
+    head_loss_factor: '0.00338'
```
Comment (haukb) [inline tests/test_unit/test_resync/approval_test/test_apply/demo_ProductionModel.yml:190]: Why is the head loss factor changing?

File: tests/test_unit/test_resync/approval_test/test_apply/demo_ProductionModel.yml [hunk 4]
```diff
@@ -115,8 +115,9 @@ assets:
   labels:
   - externalId: plant
   metadata:
+    connection_losses: '0.008919999999999999'
```
Comment (haukb) [inline tests/test_unit/test_resync/approval_test/test_apply/demo_ProductionModel.yml:118]: Weird number(?)

File: cognite/powerops/resync/config/production/plant.py [hunk 5]
```diff
@@ -64,6 +65,7 @@ def from_cdf_resources(cls, asset: Asset, relationships: list[Relationship], **k
             p_min=float(asset.metadata["p_min"]),
             p_max=float(asset.metadata["p_max"]),
             head_loss_factor=float(asset.metadata["head_loss_factor"]),
+            connection_losses=json.loads(asset.metadata.get("connection_losses") or "{}"),
```
Comment (haukb) [inline cognite/powerops/resync/config/production/plant.py:68]: Hmm, what happens here? Json.loads- but shouldn't connection_losses just be a float? Does it make sense to have this structure then? 

E.g. for head_loss_factor is loads a float only.

File: cognite/powerops/resync/models/_shared_v1_v2/production_model.py [hunk 6]
```diff
@@ -86,31 +84,70 @@ def _plant_to_inlet_reservoir_breadth_first_search(
...
         The name of the reservoir connected to the plant, or None if no reservoir was found.
     """
+
+    def get_connection_path_from_last_visited(visited_paths: list, last_visited_id: int) -> list:
+        for connection in visited_paths:
+            if connection[-1] == last_visited_id:
+                connection_path_index = visited_paths.index(connection)
+                return visited_paths[connection_path_index]
+
+    def calculate_losses_from_conncetion_path(all_junctions, all_tunnels, connection_by_id, connection_path):
```
Comment (haukb) [inline cognite/powerops/resync/models/_shared_v1_v2/production_model.py]: typo connection

Comment (haukb) [inline cognite/powerops/resync/models/_shared_v1_v2/production_model.py]: add type hints maybe?

File: cognite/powerops/resync/models/_shared_v1_v2/production_model.py [hunk 7]
```diff
@@ -86,31 +84,70 @@ def _plant_to_inlet_reservoir_breadth_first_search(
...
+        sum_losses = 0
+        order_to_loss_factor_key = {0: "loss_factor_1", 1: "loss_factor_2"}
```
Comment (haukb) [inline cognite/powerops/resync/models/_shared_v1_v2/production_model.py:96]: This seems a bit hacky, are we sure Hev other customers model it in the same way, with the same name? Maybe keep for now, but should consider whether this is a robust solution

File: cognite/powerops/resync/models/_shared_v1_v2/production_model.py [hunk 8]
```diff
@@ -86,31 +84,70 @@ def _plant_to_inlet_reservoir_breadth_first_search(
...
+            if connection.get("to_type") == "junction":
+                if connection.get("order") in order_to_loss_factor_key:
+                    junction_name = connection["to"]
+                    junction_losses = all_junctions[junction_name]
+                    loss_order = connection["order"]
+                    sum_losses += junction_losses[order_to_loss_factor_key[loss_order]]
+            elif connection.get("to_type") == "tunnel":
+                tunnel_name = connection["to"]
+                tunnel_loss = all_tunnels[tunnel_name]["loss_factor"]
+                sum_losses += tunnel_loss
```
Comment (haukb) [inline cognite/powerops/resync/models/_shared_v1_v2/production_model.py:108]: Does this function only work if we have the "to_type" concept? I know that is not the case for some SHOP model files

File: cognite/powerops/resync/models/_shared_v1_v2/production_model.py [hunk 9]
```diff
@@ -86,31 +84,70 @@ def _plant_to_inlet_reservoir_breadth_first_search(
...
+        return sum_losses
+
     queue = []
-    for connection in all_connections:
+    connection_by_id = dict(enumerate(all_connections))
```
Comment (haukb) [inline cognite/powerops/resync/models/_shared_v1_v2/production_model.py:112]: Nice touch, haven't seen this before 👍

File: cognite/powerops/resync/models/_shared_v1_v2/production_model.py [hunk 10]
```diff
@@ -86,31 +84,70 @@ def _plant_to_inlet_reservoir_breadth_first_search(
     Optional[str]
         The name of the reservoir connected to the plant, or None if no reservoir was found.
     """
+
+    def get_connection_path_from_last_visited(visited_paths: list, last_visited_id: int) -> list:
```
Comment (haukb) [inline cognite/powerops/resync/models/_shared_v1_v2/production_model.py:88]: Add docstring describing what this function does

File: cognite/powerops/resync/models/v1/config_to_model/to_production_model.py [hunk 11]
```diff
@@ -159,16 +159,23 @@ def to_production_model(configuration: config.ProductionConfig) -> ProductionMod
...
                 name=plant_name,
                 display_name=display_name,
                 ordering=str(order),
                 outlet_level=float(attributes.get("outlet_line", 0)),
                 p_min=float(attributes.get("p_min", p_min_fallback)),
                 p_max=float(attributes.get("p_max", p_max_fallback)),
-                head_loss_factor=float(
-                    attributes.get("main_loss", [head_loss_factor_fallback])[0]
-                ),  # For some reason, SHOP defines this as a list, but we only need the first (and only) value
+                head_loss_factor=sum_plant_losses,
```
Comment (haukb) [inline cognite/powerops/resync/models/v1/config_to_model/to_production_model.py:177]: Okay, I think I see the reason the head_loss_factor changes now. But is this a good idea? i.e., to add the connection losses to the main head loss factor **and** add the connection losses as a separate loss? I would strongly suggest not to to it.

File: tests/test_unit/test_resync/approval_test/test_apply/demo_ProductionModel.yml (hunk 3 above)
Comment (haukb) [reply tests/test_unit/test_resync/approval_test/test_apply/demo_ProductionModel.yml:190]: See other comment about combining connection losses to main head loss

**Review by haukb — CHANGES_REQUESTED** [review]
Brief overview with some comments. Did not quite grasp if the new approach supports having model files without "from_type" and "to_type".

File: cognite/powerops/resync/models/v1/config_to_model/to_production_model.py (hunk 11 above)
Comment (Juliamg) [reply cognite/powerops/resync/models/v1/config_to_model/to_production_model.py:177]: The reason for adding connection_losses as a metadata field as well was by request from Øicind cause he wanted a way of distinguishing what plants had connection losses and what plants did not. And I thought we agreed from the meeting with them that we should add the losses to the head_loss_factor? Or do you suggest not to sum them and have the function/consumer sum the two losses so that it is not "implicit" that head_loss_factors contains the sum

Comment (haukb) [reply cognite/powerops/resync/models/v1/config_to_model/to_production_model.py:177]: Yes. I would avoid having it implicit. 

I agreed the original plan was to add it direct to the "main head loss", but that was also when we did not plan to add the connection losses metadata field. At least that's what I thought?

Comment (Juliamg) [reply cognite/powerops/resync/models/v1/config_to_model/to_production_model.py:177]: Ok, should I remove the summation then? and then we do the sum in the function later on?

File: cognite/powerops/resync/models/_shared_v1_v2/production_model.py [hunk 12]
```diff
@@ -86,25 +92,47 @@ def _plant_to_inlet_reservoir_with_losses(
...
-                    junction_name = connection["to"]
-                    junction_losses = all_junctions[junction_name]
+                    junction_losses = all_junctions[connection_name]
                     loss_order = connection["order"]
                     sum_losses += junction_losses[order_to_loss_factor_key[loss_order]]
-            elif connection.get("to_type") == "tunnel":
-                tunnel_name = connection["to"]
-                tunnel_loss = all_tunnels[tunnel_name]["loss_factor"]
+            elif connection_name in all_tunnels:
+                tunnel_loss = all_tunnels[connection_name]["loss_factor"]
```
Comment (haukb) [inline cognite/powerops/resync/models/_shared_v1_v2/production_model.py:135]: Nice! This seems like a "safer" way to do it, although it might be a bit slower 👍

**Review by haukb — APPROVED** [review]
lgtm great work!
